src/Parsing/Impl/Definitions/IScopeNode.cs

```diff
@@ -0,0 +1,48 @@
+using Microsoft.Python.Parsing.Ast;
+
+namespace Microsoft.Python.Parsing {
+    /// <summary>
+    /// Represents a <see cref="Node"/> with a Scope
+    /// </summary>
+    public interface IScopeNode : INode {
+        /// <summary>
+        /// Scope node name
+        /// </summary>
+        string Name { get; }
+
+        /// <summary>
+        /// Parent of the current scope
+        /// </summary>
+        IScopeNode ParentNode { get; set; }
+
+        /// <summary>
+        /// Body of the scope node, null if not applicable 
+        /// </summary>
+        Statement Body { get; }
+
+        /// <summary>
+        /// Gloabl scope
+        /// </summary>
+        PythonAst GlobalParent { get; }
+
+        /// <summary>
+        /// Holds information about the scope gathered from Parser, mainly used in  <see cref="PythonNameBinder"/>
+        /// </summary>
+        ScopeInfo ScopeInfo { get; }
+
+        /// <summary>
+        /// Binds the current scope, giving <param name="binder">binder</param> access to stored variables in <see cref="ScopeInfo"/>
+        /// </summary>
+        void Bind(PythonNameBinder binder);
```

Comment:
`Bind` and `FinishBind` should not be on a public interface

Reply:
Should I make them extension methods?

Reply:
The comment here is that they have side effects, so having them be a part of a public interface is a bad idea (when they're only used internally by the parser).

Reply:
I see, so how would I make it so that it can only be used internally by the parser? I don't think I can make the interface internal. I guess I could have them as internal methods on ScopeInfo and have callers use that instead, would that be ok? Also I may have asked this before but do you agree on the idea of `ScopeInfo` or feel that I should extract the members to the `IScopeNode` interface?

Reply:
Another benefit of having `ScopeInfo` is that I can leave it in the interface and define internal methods on it so users of our assemblies cannot access them. If I had them on the interface then everyone would have access to those methods only used for the parser (assuming that there is not some way to have internal methods on a public interface)

Reply:
Internally they should be used via the class itself. It is possible to have separate internal interface too, we have quite a few. ScopeInfo technically should be `IScopeInfo` as well... But I don't like much separate 'info'. Separate entity makes sense if it is passed around as a separate object or used in other places separately. If it is only used as part of an interface, then just promote methods up to the interface. Occam Razor :-)

Reply:
Yeah, I wouldn't leave anything in the interface if it's not usable. Having something that's completely impossible to construct is a weird interface. If binding is something that's required for the parser to work but isn't generally usable outside (I think that's how it is), then it'd be better to have some separate internal interface or something.

Reply:
I see I think I'll add an internal interface then, perhaps `IBindable`. I'll try moving all the members up to the interface and moving the internal methods to the internal interface but it will take me some more time. Thanks for the help.

Reply:
I'm a little worried about code duplication however. Comprehensions should be treated like functions in terms of scoping, but without composition of their behavior then Comprehension will have to implement the `bind` and `finishBind` methods as well with the same code, do you have any suggestions for this?

Reply:
I don't think we're suggesting duplicating the code all over the place. It's fine internally for the nodes to store some common thing which implements binding, but then expose that info through a common interface (e.g. IScopeNode).

Reply:
Ok now I understand. The issue is having ScopeInfo on a public interface, not the existence of ScopeInfo itsel. ScopeInfo can be part of the implementation details and its information can be exposed through the IBindable and IScopeNode interfaces. 

I’ll have to make FunctionDefinition, ClassDefinition and the like implement another internal interface because IScopeNode cannot implement an internal interface because it is public. 

I’ll get on it though thanks for the advice.